pennylane/registers.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :func:`registers` function.
+"""
+
+from .wires import Wires
+
+
+def registers(register_dict):
+    """
+    The ``qml.registers()`` function creates a dictionary that maps register name to :class:`~.Wires`
+    object.
+
+    The function takes in a dictionary. The keys for this input dictionary are the names of the
+    registers. The values for this dictionary are either integers that represent the number of
+    wires or dictionaries that represent sub registers. For example, the key-value pair
+    `"ancilla": 3` would represent a register named "ancilla" with 3 wires whereas the key-value
+    pair `"ancilla": {"sub_ancilla": 2, "sub_ancilla1": 1}` would represent a register named
+    "ancilla" with two sub_registers: "sub_ancilla" and "sub_ancilla1", each with their respective
+    number of wires.
+
+    Given input `{"ancilla": {"sub_ancilla": 2, "sub_ancilla1": 1}}`, ``qml.registers()`` creates
+    a dictionary with 3 key-value pairs. The keys are the register names found in the input:
+    "ancilla", "sub_ancilla" and "sub_ancilla1". The values are the respective :class:`~.Wires`
+    objects for each register. For example, "ancilla" has two sub registers "sub_ancilla" and
+    "sub_ancilla1". Therefore, the value associated with key "ancilla" is the union of the
+    :class:`~.Wires` of its sub registers. Since its sub registers "sub_ancilla" and "sub_ancilla1"
+    have :class:`~.Wires` objects `Wires([0, 1])` and Wires([2]) respectively, the key "ancilla"
+    has the value `Wires([0, 1, 2])`.
```

Comment:
Thanks for the update here! Based on [this comment](https://github.com/PennyLaneAI/pennylane/pull/5957#discussion_r1688052765), I still feel like the description is a little too focused on what the function does programmatically, rather than _why_ someone would use the function.

The programmatic details can be tackled in the `Args:`, `Returns:` and example sections.

Reply:
@austingmhuang You might bring the current description to the Example section.

Reply:
I moved the programmatic description into the example sections and added back the context of why this function is useful and when to use it.